### PR TITLE
Profiler backend

### PR DESCRIPTION
**Pull request by jw3**
## Experimental profiler backend

This is a first pass at backend functionality with initial bindings.  Future PR(s) will expand the bindings and integrate with the Python `faprofiler`.

### todo
- [x] does the profiler live in its own crate?
- [x] review the `wait_for_daemon` functions in daemon and pyo3 crates and eliminate one if possible
- [x] add remaining clap arguments 

### Goals
1. Provide Rust backend that matches the profiling APIs defined by `fapd_manager.py` and `faprofiler.py`
2. Shift the profiling flow control backwards into Rust.
3. Provide a standalone executable that can profile without the GUI
4. Move  the profiler control into systemd
5. Stateless lock-free control
6. Provide bindings that allow the UI to be unchanged 
7. Allow non-deployed rules to be profiled

### Accomplished
1. > Provide Rust backend that matches the profiling APIs defined by `fapd_manager.py` and `faprofiler.py`
    - Rust backend :heavy_check_mark: 
    - Matching APIs - Follow up PRs
2. > Shift the profiling flow control backwards into Rust.
    - `daemon::Profiler` :heavy_check_mark: 
3. > Provide a standalone executable that can profile without the GUI
    - Added `faprofiler` as a CLI tool :heavy_check_mark: 
4. > Move  the profiler control into systemd
    - :heavy_check_mark: 
6. > Stateless lock-free control
    - Stateless: there is state to maintain when supporting optional features
    - lock-free: client concern
8. > Provide bindings that allow the UI to be unchanged
    - Follow up PRs
9. > Allow non-deployed rules to be profiled
    - CLI :heavy_check_mark:
    - GUI: will need some additional support

### Impacts
- #508 
- #531

### todo
- Determine if the target should be a systemd unit

### CODE REVIEW

**Comment by jw3** [comment]
This flow currently works through an on-demand system drop-in unit

```rust
    let mut profiler = Profiler::new();
    profiler.activate()?;
    let out = Command::new("ls").arg("/tmp").output()?;
    println!("{}", String::from_utf8(out.stdout)?);
    profiler.deactivate()?;
```

**Comment by tparchambault** [comment]
I'm good with your goals and your approach. Two comments

> Stateless lock-free control

We have to handle system state at some level such that two fapd instances are not running simultaneously. If we are using service units maybe the `Conflicts=xxx` directive will address this.

> Determine if the target should be a systemd unit

Even if we containerize the stand-alone, there's no reason that it can't be controlled as a systemd service.

Re: `drop-in` terminology:
> ... an on-demand system drop-in unit

We may have differing ideas of `drop-in` and my idea of it may be out-dated because I'm looking at manpages associated with an older systemd release, and systemd may have 'evolved.'

I took a quick look at your code and got the idea that you are dynamically generating the service unit file. No issues with that approach btw.

I'm trying to use the systemd meaning of the term although it's not particularly intuitive to my own personal idea of what a `drop-in` would be, but they do explicitly specify the mechanism. In any event, from `man systemd.unit`:

>  Along with a unit file foo.service, a "drop-in" directory
       foo.service.d/ may exist. All files with the suffix ".conf" from this
       directory will be parsed after the file itself is parsed. This is
       useful to alter or add configuration settings for a unit, without
       having to modify unit files. Each drop-in file must have appropriate
       section headers. Note that for instantiated units, this logic will
       first look for the instance ".d/" subdirectory and read its ".conf"
       files, followed by the template ".d/" subdirectory and the ".conf"
       files there.

So basically magic sub-directories for overrides. We could use them to override and/or extend the declarations in the shipped fapolicyd.service file to specify different stdout/stderr locations, add the --debug and --permission CL args, etc, but I'm not sure that's what you were thinking wrt `drop-ins` _or_ I could have missed something and that's exactly what you were thinking.


I'm good with all of your ideas. I like consistent APIs and control flow between similar functional sub-systems. I'll look at the code in more detail later today.

**Comment by jw3** [comment]
>> Stateless lock-free control

> We have to handle system state at some level such that two fapd instances are not running simultaneously. If we are using service units maybe the `Conflicts=xxx` directive will address this.

The goal is that checking `is_active` on a handle is enough.

>> Determine if the target should be a systemd unit

> Even if we containerize the stand-alone, there's no reason that it can't be controlled as a systemd service.

Concur.  The initial impl here is going to remain Console, but that could be changed to a `one-shot` unit that gets cleaned up at the end of the run.

> Re: `drop-in` terminology:

Your definition is accurate.  My original intent was instead of writing a service file to add drop-ins on fapolicyd, but writing a separate service file ( `fapolicyp` :stuck_out_tongue: )  is a more reliable solution to determine what is running on the system without relying on maintaining state in the application.

> I'm good with all of your ideas.

:+1: 

Now that 0.6.0 is branched off this is going to get cleaned up and merged to master.  It doesn't have to be the backend yet, but I don't want it to get stale.  I also want the cli tool to play with regularly.
